chore: support optimistic concurrency control (#3741)
FuelLabs/fuels-ts: apps/docs/src/guide/provider/rpc-consistency.md

# RPC Consistency

> A common issue with querying distributed networks is ensuring each node is consistent. At any moment each node could be at a different block height, therefore querying the state of the blockchain may yield different results from node to node.

Reviewer: Nit.

```suggestion
A common issue with querying distributed networks is ensuring each node is consistent. At any moment, each node could be at a different block height; therefore, querying the state of the blockchain may yield different results from node to node.
```

> A common issue with querying distributed networks is ensuring each node is consistent. At any moment, each node could be at a different block height; therefore, querying the state of the blockchain may yield different results from node to node.

Reviewer: ```suggestion
A common issue with querying distributed networks is ensuring consistency between nodes. At any moment, each node could be at a different block height; therefore, querying the state of the blockchain may yield different results from node to node.
```

> To defend against this, the SDK appends the block height to block-sensitive requests. The node will then ensure it has met the block height criteria before it processes the request.

Reviewer: Nit.

```suggestion
To defend against this, the SDK appends the block height to block-sensitive requests so the node will verify that it meets the block height criteria before processing the request.
```

> If the node has not met the block height criteria, the request will either be awaited on the node's side (if the node supports this) or retried from the SDK side. The SDK will attempt to retry the request until it hits maximum retries and then throw an error.

Reviewer: ```suggestion
If the node has not met the block height criteria, the request will either be awaited on the node's side (if the node supports this) or retried from the SDK side. The SDK will attempt to retry the request until it reaches the maximum number of retries and then throw an error.
```

> If the block height criteria has not been met by the node, the request will either be awaited on the node's side (if the node supports this), or it'll be retried from the SDK side. The SDK will attempt to retry the request until it hits a maximum amount of retries, and then will throw an error.

Reviewer: Nit.

```suggestion
If the node has not met the block height criteria, the request will either be awaited on the node's side (if the node supports this) or retried from the SDK side. The SDK will attempt to retry the request until it hits maximum retries and then throw an error.
```

> If the block height criteria has not been met by the node, the request will either be awaited on the node's side (if the node supports this), or it'll be retried from the SDK side. The SDK will attempt to retry the request until it hits a maximum amount of retries, and then will throw an error.

Reviewer 2: Should we include information here about the values used by the TS SDK for retry attempts?

```ts
maxRetries: 5,
baseDelay: 500,
```

Also, it would be helpful to mention that `Provider.clearChainAndNodeCaches` resets the `currentBlockHeightCache` for all provider instances associated with the given URL or for all URLs when a specific URL is not informed.

On a related note, we should also clarify that all Provider instances created with the same URL share the same `currentBlockHeightCache`. Including code snippets with multiple provider instances could help make this behavior easier to understand.

Author: > Should we include information here about the values used by the TS SDK for retry attempts?

Yes, it might be useful for users to make a more informed decision on if they want the flag enabled or not.

> Also, it would be helpful to mention that `Provider.clearChainAndNodeCaches` resets the `currentBlockHeightCache` for all provider instances associated with the given URL or for all URLs when a specific URL is not informed.

IMO it's best to document that on the `clearChainAndNodeCaches` method itself.

> On a related note, we should also clarify that all Provider instances created with the same URL share the same `currentBlockHeightCache`. Including code snippets with multiple provider instances could help make this behavior easier to understand.

I didn't find any documentation generally around the `Provider` having a static instance of the cache that is accessed by all `Provider` instances. This should be documented not only for `currentBlockHeightCache` but also for the other caches. I think this would be better done in a subsequent PR that's focused on documenting it all, `currentBlockHeightCache` included.

> If the block height criteria has not been met by the node, the request will either be awaited on the node's side (if the node supports this), or it'll be retried from the SDK side. The SDK will attempt to retry the request until it hits a maximum amount of retries, and then will throw an error.

Reviewer 2: Should we include information here about the values used by the TS SDK for retry attempts?

```ts
maxRetries: 5,
baseDelay: 500,
```

Also, it would be helpful to mention that `Provider.clearChainAndNodeCaches` resets the `currentBlockHeightCache` for all provider instances associated with the given URL or for all URLs when a specific URL is not informed.

On a related note, we should also clarify that all Provider instances created with the same URL share the same `currentBlockHeightCache`. Including code snippets with multiple provider instances could help make this behavior easier to understand.

Reviewer 2: Duplicated

> This functionality is enabled by default, but can be disabled as so:

Reviewer: ```suggestion
This functionality is enabled by default but can be disabled as so:
```

# RPC Consistency

> Because the Fuel network consists of numerous nodes that at the same moment might be at different block heights, querying the state of the blockchain might yield different results from node to node. To fix that, every node appends its current block height to every response it returns. This enables the SDK to then append this block height to block-sensitive requests that it makes towards the network, thereby requiring the node it hits to be at that block height before processing the request.
> If the node is at a lower block height, the request will either be awaited on the node's side if the node is configured to do so, or it'll be retried from the SDK side for a couple of times with an exponential backoff. Once the maximum retry amount is reached, the SDK will throw an error.
> 
> This functionality is turned on by default. If you want to turn it off to e.g. have faster response times, you can do the following:

Reviewer 3: ```suggestion
A common issue with querying distributed networks is ensuring each node is consistent. At any moment each node could be at a different block height, therefore querying the state of the blockchain may yield different results from node to node.

To defend against this, the SDK appends the block height to block-sensitive requests. The node will then ensure it has met the block height criteria before it processes the request.

If the block height criteria has not been met by the node, the request will either be awaited on the node's side (if the node supports this), or it'll be retired from the SDK side. The SDK will attempt to retry the request until it hits a maximum amount of retries, and then will throw an error.

This functionality is enabled by default, but can be disabled as so:
```

Rewritten to break it up a bit and improve the flow

> This functionality is turned on by default. If you want to turn it off to e.g. have faster response times, you can do the following:

Author: I couldn't say for certain; it depends on the speed of block syncing and how much the node we hit lags behind.
